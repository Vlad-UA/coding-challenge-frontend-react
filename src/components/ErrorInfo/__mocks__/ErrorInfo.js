// Core
import React from 'react';

const ErrorInfoMock = () => <div className="error-info-mock">ErrorInfoMock</div>;
ErrorInfoMock.displayName = 'ErrorInfoMock';
export default ErrorInfoMock;
