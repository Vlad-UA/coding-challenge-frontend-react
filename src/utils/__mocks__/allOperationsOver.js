export const allOperationsOver = () => new Promise(resolve => setTimeout(resolve, 0));
