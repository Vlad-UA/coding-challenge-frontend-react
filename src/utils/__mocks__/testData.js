const OCCURRED_AT = 1552325609;

class IncidentData {
  constructor(
    id = Math.floor(10000 + Math.random() * 90000),
    title = `Title ${id}`,
    description = `Description ${id}`,
    address = `Address ${id}`,
    occurred_at = OCCURRED_AT,
    updated_at = 1553542734,
    image_url_thumb = 'https://files.bikeindex.org/uploads/Pu/148912/small_IMG_20170705_191324.jpg',
    type = 'Theft',
  ) {
    this.id = id;
    this.title = title;
    this.description = description;
    this.address = address;
    this.occurred_at = occurred_at;
    this.updated_at = updated_at;
    this.media = {
      image_url_thumb,
    };
    this.type = type;
  }
}

// One Incident
export class IncidentModel {
  constructor({
    id,
    title,
    description,
    address,
    occurred_at,
    updated_at,
    image_url_thumb,
    type,
  }) {
    this.incident = new IncidentData(
      id,
      title,
      description,
      address,
      occurred_at,
      updated_at,
      image_url_thumb,
      type,
    );
  }
}

// Many Incidents
export class IncidentsModel {
  constructor(
    ids = [97572, 94547],
  ) {
    this.incidents = ids.map(id => new IncidentData(id));
  }
}


class LocationData {
  constructor(
    id = Math.floor(10000 + Math.random() * 90000),
    occurred_at = OCCURRED_AT,
  ) {
    this.properties = {
      id,
      occurred_at,
    };
    this.geometry = {
      type: 'Point',
      coordinates: [
        -116.1960257,
        43.6334557,
      ],
    };
  }
}

export class LocationsModel {
  constructor(
    ids = [97572],
  ) {
    this.type = 'FeatureCollection';
    this.features = ids.map(id => new LocationData(id));
  }
}
