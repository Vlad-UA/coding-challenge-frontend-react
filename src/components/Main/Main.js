// Core
import React, { PureComponent } from 'react';

// Instruments
import styled from 'styled-components';

// Utils
import callApi from '../../utils/call-api';

// Components
import SearchPanel from '../SearchPanel/SearchPanel';
import ErrorInfo from '../ErrorInfo/ErrorInfo';
import IncidentList from '../IncidentList/IncidentList';
import Header from '../Header/Header';

const Wrapper = styled.div`margin: 50px;`;

export default class Main extends PureComponent {
  state = {
    isDataError: false,
    isDataLoading: false,
    incidents: [],
    currentPage: 1,
  };

  async componentDidMount() {
    this.isComponentMounted = true;

    await this.getData();
  }

  componentWillUnmount() {
    this.isComponentMounted = false;
  }

  getData = async ({ dateFrom, dateTo, query } = {}) => {
    this.setState({
      isDataError: false, isDataLoading: true,
    });

    try {
      const { incidents } = await callApi(
        { occurredAfter: dateFrom, occurredBefore: dateTo, query },
      );
      if (this.isComponentMounted) {
        this.setState({ incidents });
      }
    } catch (error) {
      if (this.isComponentMounted) {
        this.setState({ isDataError: true });
      }
    } finally {
      if (this.isComponentMounted) {
        this.setState({ isDataLoading: false });
      }
    }
  };

  onFind = async ({ dateFrom, dateTo, query }) => {
    this.setState({ currentPage: 1 });
    await this.getData({ dateFrom, dateTo, query });
  };

  openPage = (pageNumber) => {
    this.setState({ currentPage: pageNumber });
    window.scrollTo(0, 0);
  };

  dataLayout = () => {
    const {
      isDataError, incidents, currentPage, isDataLoading,
    } = this.state;

    if (isDataError) {
      return <ErrorInfo />;
    } if (isDataLoading) {
      return <p>Loading ... </p>;
    }
    return (
      <IncidentList
        incidents={incidents}
        currentPage={currentPage}
        openPage={this.openPage}
      />
    );
  };

  render() {
    const { isDataLoading } = this.state;

    return (
      <Wrapper>
        <Header />

        <SearchPanel onFind={this.onFind} isDataLoading={isDataLoading} />

        {this.dataLayout()}
        <p>v 1.0.0</p>
      </Wrapper>
    );
  }
}
