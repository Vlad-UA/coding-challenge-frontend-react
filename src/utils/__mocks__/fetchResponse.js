// Test utilities
import { IncidentsModel } from './testData';

export class FetchResponse {
  constructor({
    status = global.fetchResponseStatus,
    fetchResponseData = new IncidentsModel(),
  } = {}) {
    return {
      status,
      json() {
        return Promise.resolve(fetchResponseData);
      },
    };
  }
}
