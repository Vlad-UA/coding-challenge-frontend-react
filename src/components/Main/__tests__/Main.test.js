/* eslint-env jest */

// Core
import React from 'react';

// Test utilities
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { allOperationsOver } from '../../../utils/__mocks__/allOperationsOver';

// Components
import Main from '../Main';

jest.mock('../../ErrorInfo/ErrorInfo');

describe('Component Main:', () => {
  test('Must match snapshot with "Loading...":', () => {
    const component = shallow(<Main />);

    expect(toJSON(component)).toMatchSnapshot();
  });

  test('Must match snapshot with component "IncidentList":', async () => {
    const component = await shallow(<Main />);

    await allOperationsOver();

    expect(toJSON(component)).toMatchSnapshot();
  });


  test('Must shown the error component if error on call API', async () => {
    global.fetchResponseStatus = 404;

    const component = shallow(<Main />);
    await allOperationsOver();
    expect(component.text().includes('ErrorInfoMock')).toBe(true);

    global.fetchResponseStatus = 200;
  });
});
