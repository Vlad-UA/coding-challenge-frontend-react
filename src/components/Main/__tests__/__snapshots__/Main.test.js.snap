// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Main: Must match snapshot with "Loading...": 1`] = `
<div
  className="Main"
>
  <Header />
  <SearchPanel
    isDataLoading={true}
    onFind={[Function]}
  />
  <p>
    Loading ... 
  </p>
</div>
`;

exports[`Component Main: Must match snapshot with component "IncidentList": 1`] = `
<div
  className="Main"
>
  <Header />
  <SearchPanel
    isDataLoading={false}
    onFind={[Function]}
  />
  <IncidentList
    currentPage={1}
    incidents={
      Array [
        IncidentData {
          "address": "Address 97572",
          "description": "Description 97572",
          "id": 97572,
          "media": Object {
            "image_url_thumb": "https://files.bikeindex.org/uploads/Pu/148912/small_IMG_20170705_191324.jpg",
          },
          "occurred_at": 1552325609,
          "title": "Title 97572",
          "type": "Theft",
          "updated_at": 1553542734,
        },
        IncidentData {
          "address": "Address 94547",
          "description": "Description 94547",
          "id": 94547,
          "media": Object {
            "image_url_thumb": "https://files.bikeindex.org/uploads/Pu/148912/small_IMG_20170705_191324.jpg",
          },
          "occurred_at": 1552325609,
          "title": "Title 94547",
          "type": "Theft",
          "updated_at": 1553542734,
        },
      ]
    }
    openPage={[Function]}
  />
</div>
`;
